docs/design/caching_scan_results_by_repo_digest.md

```diff
@@ -0,0 +1,117 @@
+I1008 19:58:19.358802   62385 scanner.go:89] Checking if images were already scanned
+I1008 19:58:19.360411   62385 scanner.go:95] Cached scan reports: 0
+I1008 19:58:19.360421   62385 scanner.go:101] Scanning with options: {ScanJobTimeout:0s DeleteScanJob:true}
+I1008 19:58:19.365155   62385 runner.go:79] Running task and waiting forever
+I1008 19:58:19.365190   62385 runnable_job.go:74] Creating job "starboard/scan-vulnerabilityreport-cbf8c9b99"
+I1008 19:58:19.376902   62385 reflector.go:219] Starting reflector *v1.Event (30m0s) from pkg/mod/k8s.io/client-go@v0.22.2/tools/cache/reflector.go:167
+I1008 19:58:19.376920   62385 reflector.go:255] Listing and watching *v1.Event from pkg/mod/k8s.io/client-go@v0.22.2/tools/cache/reflector.go:167
+I1008 19:58:19.376902   62385 reflector.go:219] Starting reflector *v1.Job (30m0s) from pkg/mod/k8s.io/client-go@v0.22.2/tools/cache/reflector.go:167
+I1008 19:58:19.376937   62385 reflector.go:255] Listing and watching *v1.Job from pkg/mod/k8s.io/client-go@v0.22.2/tools/cache/reflector.go:167
+I1008 19:58:19.386049   62385 runnable_job.go:130] Event: Created pod: scan-vulnerabilityreport-cbf8c9b99-4nzkb (SuccessfulCreate)
+I1008 19:58:51.243554   62385 runnable_job.go:130] Event: Job completed (Completed)
+I1008 19:58:51.247251   62385 runnable_job.go:109] Stopping runnable job on task completion with status: Complete
+I1008 19:58:51.247273   62385 runner.go:83] Stopping runner on task completion with error: <nil>
+I1008 19:58:51.247278   62385 scanner.go:130] Scan job completed: starboard/scan-vulnerabilityreport-cbf8c9b99
+I1008 19:58:51.247297   62385 scanner.go:262] Getting logs for nginx container in job: starboard/scan-vulnerabilityreport-cbf8c9b99
+I1008 19:58:51.674449   62385 scanner.go:123] Deleting scan job: starboard/scan-vulnerabilityreport-cbf8c9b99
+```
+
+Now, if we scan the `nginx` Deployment in the `bar` namespace, Starboard will
+see that there's already a ClusterVulnerabilityReport (`84bcb5cd46`) for the
+same image reference `nginx:1.16` and will skip creation of a K8s Job. It will
+just read and copy the report as VulnerabilityReport object to the `bar`
+namespace.
+
+```console
+$ kubectl get clustervulnerabilityreports -o wide
+NAME         REPOSITORY      TAG    DIGEST   SCANNER   AGE   CRITICAL   HIGH   MEDIUM   LOW   UNKNOWN
+84bcb5cd46   library/nginx   1.16            Trivy     17s   21         50     33       104   0
+```
+
+```console
+$ starboard scan vulnerabilityreports deploy/nginx -n bar -v 3
+I1008 19:59:23.891718   62478 scanner.go:72] Getting Pod template for workload: {Deployment nginx bar}
+I1008 19:59:23.895310   62478 scanner.go:89] Checking if image nginx:1.16 was already scanned
+I1008 19:59:23.903058   62478 scanner.go:95] Cache hit
+I1008 19:59:23.903078   62478 scanner.go:97] Copying ClusterVulnerabilityReport to VulnerabilityReport
+```
+
+As you can see, Starboard eventually created two VulnerabilityReports by spinning
+up only one K8s Job.
```

Comment:
Which problem are we trying to solve: 1. Spinning less K8s jobs or 2. Creating less VulnerabilityReports?

I ask this because https://github.com/aquasecurity/starboard/pull/740/files#diff-fa88466a8291355bf7dddafd314bcf4277f669d956110f8fdfd319dd48f8147eR32-R35 mentions 2 jobs and 2 VulnerabilityReports. From this solution all we've done is 1 job and 2 VulnerabilityReports. 

If long running jobs are the problem we're trying to solve, wouldn't using Trivy Client/Server model be a better idea? (Reasons: Less code, less bugs, single cache, and ultimately faster finishing jobs).

Reply:
In this proposal we are just trying to create less K8s jobs. For VulnerabilityReports we'll have a separate design.

Starboard is already integrated with Trivy in [Standalone] and [ClientServer] mode.

By default we use Trivy in Standalone mode because it does not require deploying Trivy server. It's good for demos and trying things out, but is also least efficient.

Even if you switch from Standalone to ClientServer mode Starboard would still create K8s jobs. The only difference is that jobs will complete faster. Therefore, leveraging cache allows us to skip K8s jobs creation.

[Standalone]: https://aquasecurity.github.io/starboard/v0.12.0/integrations/vulnerability-scanners/trivy/#standalone
[ClientServer]: https://aquasecurity.github.io/starboard/v0.12.0/integrations/vulnerability-scanners/trivy/#clientserver